src/app.ts

```diff
@@ -35,23 +47,161 @@ const app = new Koa();
-      creator_uid: op.get("creator_uid"),
-      op_code: op.get("op_code"),
-      data: op.get("data")
-    })
+async function validateUid(uid, ctx, next) {
+  const uidDoc = await members.doc(uid).get();
+  if (!uidDoc.exists) {
+    ctx.status = 404;
+    return;
+  }
+  ctx.state.toMember = uidDoc;
+  return await next();
+}
+
+function getPrivateVideoRef(memberUid): Storage.File {
+  return storage
+    .bucket(PRIVATE_VIDEO_BUCKET)
+    .file(`private-video/${memberUid}/invite.mp4`);
+}
+
+function getPublicVideoRef(memberMid): Storage.File {
+  return storage.bucket(PUBLIC_VIDEO_BUCKET).file(`${memberMid}/invite.mp4`);
+}
+
+async function createCoconutVideoEncodingJob(memberUid, creatorMid) {
+  const videoRef = getPrivateVideoRef(memberUid);
+
+  const temporaryAccessUrl = (await videoRef.getSignedUrl({
+    action: "read",
+    expires: Date.now() + TEN_MINUTES
+  }))[0];
+
+  const webhookUrl = new URL(
+    `/api/members/${memberUid}/notify_video_encoded`,
+    "https://raha-5395e.appspot.com"
   );
-  ctx.body = JSON.stringify(parsedOps);
-  ctx.status = 200;
-  return;
-});
+  const uploadUrl = new URL(
+    `/api/members/${memberUid}/upload_video`,
+    "https://raha-5395e.appspot.com"
+  );
+  uploadUrl.searchParams.append("mid", creatorMid);
+
+  coconut.createJob(
+    {
+      api_key: coconut_api_key,
+      source: temporaryAccessUrl,
+      webhook: webhookUrl.toString(),
+      outputs: {
+        mp4: uploadUrl.toString()
+      }
+    },
+    job => {
+      if (job.status === "ok") {
+        console.log("Coconut encoding job created successfully.");
+        console.log(job.id);
+      } else {
+        console.error("Coconut encoding job error", job.error_code);
+        console.error(job.error_message);
+      }
+    }
+  );
+}
+
+const publicRouter = new Router()
+  .param("uid", validateUid)
+  .get("/api/operations", async ctx => {
+    // TODO: Do we need to paginate?
+    const ops = await operations.orderBy("created_at").get();
+    const parsedOps: Array<Object> = [];
+    ops.forEach(op =>
+      parsedOps.push({
+        id: op.id,
+        creator_mid: op.get("creator_mid"),
+        creator_uid: op.get("creator_uid"),
+        op_code: op.get("op_code"),
+        data: op.get("data")
+      })
+    );
+    ctx.body = JSON.stringify(parsedOps);
+    ctx.status = 200;
+    return;
+  })
+  /* These endpoints are consumed by coconut. */
+  .post("/api/members/:uid/notify_video_encoded", async ctx => {
+    try {
+      const videoRef = getPrivateVideoRef(ctx.state.toMember.id);
+      if ((await videoRef.exists())[0]) {
+        await videoRef.delete();
+      } else {
+        console.warn("We expected a private video file to exist that did not.");
+      }
+      ctx.status = 200;
+    } catch (error) {
+      console.error(error);
+      ctx.status = 500;
+    }
+    return;
+  })
+  .post("/api/members/:uid/upload_video", ctx => {
+    return new Promise(async (resolve, reject) => {
```

Comment:
This is a really weird pattern, so this is a function that returns a promise that calls a function that returns promises? And within, it will resolve even when the api requests fail?

You can use bluebird to turn callback-style functions to be promise style, and preferably any failure cases ought to reject (and any appropriate return to the user can be handled by an error-handling middleware). Though if the streams API is a little awkward to do with async/await, you can just wrap that small piece of code in a promise and do the rest async/await style.

Reply:
Ahhhh. Thank you. Yeah this was super ugly but I didn't want to deal with promisify'ing the stream. Totes didn't need to wrap the whole thing in a promise, though. Still getting use to Promises, I think XD